engine/src/asset/sound/dat/mixer.rs

```diff
@@ -1,51 +1,541 @@
+};
 
-pub struct Mixer {
-    samples: Samples,
+use itertools::Itertools;
+
+use super::{
+    finetune::FineTune, pattern_effect::*, pattern_event::*, t_instrument::*, t_song::TSong,
+};
+use crate::asset::sound::sample::{AudioBuffer, AudioSamplePoint};
+
+#[derive(Default, Clone, Debug)]
+struct PlayerChannelNote {
+    finetune: Option<FineTune>,
+    finetune_initial: Option<FineTune>,
+    on: bool,
 }
 
-impl Mixer {
-    pub fn new() -> Self {
-        Self {
-            samples: Vec::new(),
+#[derive(Default, Clone)]
+struct PlayerChannel {
+    instrument: Option<Rc<TInstrument>>,
+    sample: Option<Rc<TSample>>,
+    note: PlayerChannelNote,
+    volume: f32,
+    volume_target: f32,
+    volume_actual: f32,
+    effects: [Option<PatternEffect>; 2],
+    effects_memory: HashMap<PatternEffectMemoryKey, PatternEffect>,
+    note_delay: usize,
+    retrigger: bool,
+
+    previous: Option<(Box<Self>, f64)>,
+
+    pos_sample: f64,
+    pos_volume_envelope: usize,
+    direction: PlaybackDirection,
+}
+
+impl PlayerChannel {
+    // Length of a transition between the current and the next sample in seconds
+    // Too large of a time and samples will audibly blend and play 2 notes at the same time, which sounds weird.
+    // Too little and transitions between notes will click.
+    // Chosen value is a bit of an arbitrary value that I found sounds nice.
+    // It amounts to:
+    // - 13 samples at 16000
+    // - 35 samples at 44100
+    // - 38 samples at 48000
+    const SAMPLE_BLEND: f64 = Duration::from_micros(800).as_secs_f64();
+    // Maximum difference in volume between 2 audio samples
+    // Volume as in channels volume, does not account for samples
+    // A bit of an arbitrary amount too
+    const MAX_VOLUME_CHANGE: f32 = 1. / 128.;
+
+    fn note_cut(&mut self) {
+        self.volume = 0.;
+        self.volume_actual = 0.;
+    }
+
+    fn pos_reset(&mut self) {
+        self.pos_sample = 0.;
+        self.pos_volume_envelope = 0;
+        self.direction = PlaybackDirection::Forwards;
+    }
+
+    fn compute_volume_envelope(&self) -> Option<f32> {
+        self.instrument.as_ref().and_then(|i| match &i.volume {
+            TInstrumentVolume::Envelope(envelope) => {
+                if self.note.on {
+                    Some(
+                        envelope
+                            .volume_beginning()
+                            .get(self.pos_volume_envelope)
+                            .copied()
+                            .unwrap_or(envelope.volume_loop()),
+                    )
+                } else {
+                    envelope
+                        .volume_end()
+                        .get(
+                            self.pos_volume_envelope
+                                .saturating_sub(envelope.volume_beginning().len()),
+                        )
+                        .copied()
+                }
+            }
+            TInstrumentVolume::Constant(_) => Some(1.),
```

Comment:
The value inside `TInstrumentVolume::Constant` is always `1.0`...

```suggestion
            TInstrumentVolume::Constant(volume) => Some(volume),
```

Reply:
Yeah no clue why I did it. I refactored it to an `Option` instead.